k2/csrc/intersect_pruned.cu

```diff
@@ -0,0 +1,1044 @@
+/**
+ * @brief
+ * compose
+ *
+ * @copyright
+ * Copyright (c)  2020  Xiaomi Corporation (authors: Daniel Povey)
+ *
+ * @copyright
+ * See LICENSE for clarification regarding multiple authors
+ */
+
+#include <limits>
+#include <vector>
+
+#include "k2/csrc/array_ops.h"
+#include "k2/csrc/fsa_algo.h"
+#include "k2/csrc/fsa_utils.h"
+#include "k2/csrc/ragged_ops.h"
+
+namespace k2 {
+
+namespace intersect_pruned_internal {
+
+/* Information associated with a state active on a particular frame..  */
+struct StateInfo {
+  /* abs_state_id is the state-index in a_fsas_.  Note: the ind0 in here
+     won't necessarily match the ind0 within FrameInfo::state if
+     a_fsas_stride_ == 0. */
+  int32_t a_fsas_state_idx01;
+
+  /* Caution: this is ACTUALLY A FLOAT that has been bit-twiddled using
+     FloatToOrderedInt/OrderedIntToFloat so we can use atomic max.  It
+     represents a Viterbi-style 'forward probability'.  (Viterbi, meaning: we
+     use max not log-sum).  You can take the pruned lattice and rescore it if
+     you want log-sum.  */
+  int32_t forward_loglike;
+
+  /* Note: this `backward_loglike` is the best score of any path from here to
+     the end, minus the best path in the overall FSA, i.e. it's the backward
+     score you get if, at the final-state, you set backward_loglike ==
+     forward_loglike. So backward_loglike + OrderedIntToFloat(forward_loglike)
```

Comment:
Is it assigned to **negative** `forward_loglike`?

Reply:
yes, will fix.